text/0000-unsafe-block-in-unsafe-fn.md

```diff
@@ -0,0 +1,135 @@
+making an assumption, another party having the obligation to prove that
+assumption).  There is no reason to believe that the assumption made by the
+`unsafe fn` is the same as the obligation incurred by unsafe operations inside
+this function, and hence the author of the `unsafe fn` should better carefully
+check that their assumptions are sufficient to justify the unsafe operations
+they are performing.  This is what an `unsafe` block would indicate.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When you perform an unsafe operation, like dereferencing a raw pointer or
+calling an `unsafe` function, you must enclose that code in an `unsafe` block.
+The purpose of this is to acknowledge that the operation you are performing here
+has *not* been checked by the compiler, you are responsible yourself for
+upholding Rust's safety guarantees.  Generally, unsafe operations come with
+detailed documentation for the conditions that must be met when this operation
+is executed; it is up to you to check that all these conditions are indeed met.
+
+When you are writing a function that itself has additional conditions to ensure
+safety (say, it accesses some data without making some necessary bounds checks,
+or it takes some raw pointers as arguments and performs memory operations based
+on them), then you should mark this as an `unsafe fn` and it is up to you to
+document the conditions that must be met for the arguments.
+
+Your `unsafe fn` will likely perform unsafe operations; these have to be
+enclosed by an `unsafe` block as usual.  This is the place where you have to
+check that the requirements you documented for your own function are sufficient
+to satisfy the conditions required to perform this unsafe operation.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+First of all, we no longer warn that an `unsafe` block is unnecessary when it is
+nested immediately inside an `unsafe fn`.  So, the following compiles without
+any warning:
+
+```rust
+unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T {
+  unsafe { x.get_unchecked(i) }
+}
+```
+
+However, nested `unsafe` blocks are still redundant, so this warns:
+
+```rust
+unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T {
+  unsafe { unsafe { x.get_unchecked(i) } }
+}
+```
+
+In a next step, we have a lint that fires when an unsafe operation is performed
+inside an `unsafe fn` but outside an `unsafe` block.  So, this would trigger the
+lint:
+
+```rust
+unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T {
+  x.get_unchecked(i)
+}
+```
+
+This gets us into a state where programmers are much less likely to accidentally
+perform undesired unsafe operations inside `unsafe fn`.
+
+Even later, it might be desirable to turn this warning into an error.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This new warning will likely fire for the vast majority of `unsafe fn` out there.
+
```

Comment:
Other possible drawbacks to list:

1. It will become less ergonomic to write unsafe code (it's justified I think, but worth mentioning...).

2. People might just do this:

```rust
unsafe fn frobnicate(x: T, y: U, ...) -> R {
    unsafe {
        ... // Actual code.
    }
}
```

and then nothing has been gained. I don't know what the risk of this is, but worth mentioning.

Reply:
I added (a variant of) 1.

For 2., I think something has been gained: It is not possible to incrementally improve this function's `unsafe` locality. Or maybe it is not worth it, then that has at least been explicitly documented in the code.

Reply:
The drawbacks section now says

> Many `unsafe fn` are actually rather short (no more than 3 lines) and will
likely end up just being one large `unsafe` block.  This change would make such functions less ergonomic to write.

That should cover 2, right?

Reply:
@RalfJung the concern is actually slightly different here; it is that people will just go and write one big `unsafe { ... }` when they shouldn't.

Reply:
@mark-i-m yes; sure -- the concern is that the change we do here might not have any noticable effect cause people could be lazy and...

Reply:
Wouldn't it be good that people can explicitly choose to _not_ write a big `unsafe` block instead of being forced into one automatically?